state/sqlserver/migration.go

```diff
@@ -272,35 +271,85 @@ func (m *migration) createStoredProcedureIfNotExists(db *sql.DB, name string, es
 /* #nosec */
 func (m *migration) ensureUpsertStoredProcedureExists(db *sql.DB, mr migrationResult) error {
 	tsql := fmt.Sprintf(`
-		CREATE PROCEDURE %s (
-			@Key 			%s,
-			@Data 			NVARCHAR(MAX),
-			@RowVersion 	BINARY(8))
-		AS
-			IF (@RowVersion IS NOT NULL)
-			BEGIN
-				UPDATE [%s]
-				SET [Data]=@Data, UpdateDate=GETDATE()
-				WHERE [Key]=@Key AND RowVersion = @RowVersion
-
-				RETURN
-			END
-
-			BEGIN TRY
-				INSERT INTO [%s] ([Key], [Data]) VALUES (@Key, @Data);
-			END TRY
-
-			BEGIN CATCH
-				IF ERROR_NUMBER() IN (2601, 2627)
-				UPDATE [%s]
-				SET [Data]=@Data, UpdateDate=GETDATE()
-				WHERE [Key]=@Key AND RowVersion = ISNULL(@RowVersion, RowVersion)
-			END CATCH`,
+			CREATE PROCEDURE %s (
```

Comment:
Overall, I think the logic here is still problematic. I understand the desire not to churn existing code, but it looks like there are still issues relative to the expected behaviors expressed in #2739 1c. Condensed here for reference:

StateOptions.Concurrency | Req.Etag | Expected Behavior in State Store
-- | -- | --
CONCURRENCY_FIRST_WRITE | not nil | Write if etag matches
CONCURRENCY_FIRST_WRITE | nil | Write if key does not exist
CONCURRENCY_LAST_WRITE or unspecified | not nil | Return error (Upcoming breaking change. Today, state components write regardless of etag)
CONCURRENCY_LAST_WRITE or unspecified | nil | Write regardless of etag

"Write" in this case refers to an upsert (i.e. insert key if it does not already exist, otherwise update). The original code ignores the top half of that table, but it also looks like it already did not implement the bottom half correctly. Please correct any readings of the code here: 

- If an ETag is provided and the key does not exist, it attempts to update, which returns 0 rows modified but does not error and does not insert the value. It is silently lost.
- If an ETag is provided and the key does exist, it will attempt to update using the ETag. If the ETag does not match, it again does not error and does not update the value, which has FIRST_WRITE semantics instead of LAST_WRITE (default) semantics. It should ignore the ETag completely in this case.

Preserving the original logic propagates these issues forwards. The new logic for handling FIRST_WRITE works, although it raises some questions about the specification for behavior:

- If ETag provided and key does not exist, ETag can never match. Should the method raise an error or silently ignore the write? 
    - This implementation returns an error
- If ETag provided and key exists, but ETag does not match, should the method raise an error or silently ignore the write?
    - This implementation silently ignores the update. (The update on row 290 finds no matches).  

That seems potentially inconsistent as an error contract for Dapr.

@artursouza To keep this moving forwards, since you've already approved the PR and the new code is okay, I suggest we file a separate issue to track the logical inconsistencies in the existing code. In continuing the discussion of Dapr ETag/concurrency state handling, I'll also follow up on the issue with my questions on the expected error handling.

Reply:
The issue for LAST_WRITE and eTag not null is behavior change beyond the scope of this change because it is a broader concern of the expectations for the state API behavior. I am OK with the current logic in this component.

The error handling for silent etag mistmatch in the stored procedure is handled in code by checking number of rows affected.

@CodeMonkeyLeet Let me know if you have any other concerns on this PR.

Reply:
Nope, I think it's definitely a step forward and we should take it, thanks for reminding me about the downstream handling of returning errors from the zero rows updated check.